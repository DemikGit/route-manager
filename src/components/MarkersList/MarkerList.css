.marker-list__container {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
}

.marker-list__field {
  margin-top: 5px;
  margin-bottom: 5px;
}
