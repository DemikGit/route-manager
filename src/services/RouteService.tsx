import * as React from 'react';
import { GoogleMap, InfoWindow, Marker, Polyline } from 'react-google-maps';

export interface IMarker {
  id: string,
  name: string,
  draggable: boolean,
  showInfoBox: boolean,
  onDrag: (event: google.maps.MouseEvent, id: string) => void,
  position: {
    lat: number,
    lng: number,
  }
}

export class RouteService {
  /* tslint:disable variable-name*/
  private _markers: IMarker[];
  private _mapRef: React.RefObject<GoogleMap>;
  private _componentUpdate: () => void;
  /* tslint:enable variable-name */

  private defaultPoint: IMarker;
  private polylineOptions: {
      strokeColor: string,
      strokeOpacity: number,
      strokeWeight: number,
  };

  public set updateComponent(componentUpdate: () => void) {
    this._componentUpdate = componentUpdate;
  }

  public get markers() : IMarker[] {
    return this._markers;
  }

  public set mapRef(mapRef: React.RefObject<GoogleMap>) {
    this._mapRef = mapRef;
  }

  constructor() {
    this._markers = [];
    this.polylineOptions = {
      strokeColor: '#4bdded',
      strokeOpacity: 0.85,
      strokeWeight: 4,
    };
    this.defaultPoint = {
      draggable: true,
      id: '',
      name: '',
      onDrag: (event: google.maps.MouseEvent, id: string) => {
        this.onDrag(event, id);
      },
      position: {
        lat: 0,
        lng: 0,
      },
      showInfoBox: false,
    };
  }

  public swapPoints = (srcId: string, targetId: string): void => {
    if(srcId !== targetId) {
      const newMarkers = this._markers;

      const srcMarkerIndex = this._markers.findIndex((marker: IMarker) => {
        return marker.id === srcId;
      });
      const srcMarker = this._markers[srcMarkerIndex];

      const targetMarkerIndex = this._markers.findIndex((marker: IMarker) => {
        return marker.id === targetId;
      });
      const targetMarker = this._markers[targetMarkerIndex];

      newMarkers[targetMarkerIndex] = srcMarker;
      newMarkers[srcMarkerIndex] = targetMarker;

      this._markers = newMarkers;
      this._componentUpdate();
    }
  }

  public onDrag = (event: google.maps.MouseEvent, id: string): void => {
    const { lat, lng } = event.latLng;

    const markerTarget: IMarker | undefined =
      this._markers.find((point: IMarker) => {
        return point.id.includes(id);
    });

    if (markerTarget) {
      markerTarget.position = { lat: lat() , lng: lng() };
    }

    this._componentUpdate();
  }

  public onPointDelete = (id: string): void => {
    this._markers = this._markers.filter((point: IMarker) => {
      return !point.id.includes(id);
    });

    this._componentUpdate();
  }

  public onAddPoint = (name: string): void => {
    if (this._mapRef.current) {

      const {
        lat: centerLat,
        lng: centerLng,
      } = this._mapRef.current.getCenter();

      this._markers.push({
        ...this.defaultPoint,
        id: this.uuidV4(),
        name,
        position: { lat: centerLat(), lng: centerLng() },
      });

      this._componentUpdate();
    }
  }

  public render = (): JSX.Element[] => {
    return this.renderMarkers().concat(this.renderPolyline());
  }

  private onToggleInfoBox = (id: string): void => {
    const targetMarker: IMarker | undefined = this._markers.find(
      (marker: IMarker): boolean => {
        return marker.id === id;
      }
    );

    if( targetMarker ) {
      targetMarker.showInfoBox = !targetMarker.showInfoBox;
    }

    this._componentUpdate();
  }

  private renderMarkers = (): JSX.Element[] => {
    const markers: JSX.Element[] = [];

    this._markers.forEach((marker) => {
      const { onDrag, draggable, position, id } = marker;
      const extendedDrag = ( event: google.maps.MouseEvent) => {
        onDrag(event, id);
      };

      const extendedToggleInfoBox = ( ) => {
        this.onToggleInfoBox(id);
      };

      markers.push(
        <Marker
          key={ id }
          onClick={ extendedToggleInfoBox }
          onDrag={ extendedDrag }
          draggable={ draggable }
          position={ position }
          labelAnchor={ new google.maps.Point(0, 0) }
        >
          { marker.showInfoBox && (
            <InfoWindow
              onCloseClick={ extendedToggleInfoBox }
            >
              <div>
                {`  ${ marker.name }`}
              </div>
            </InfoWindow>
          )}
        </Marker>
      );
    })

    return markers;
  }

  private renderPolyline = (): JSX.Element[] => {
    if (this._markers.length > 0) {
      const path: google.maps.LatLngLiteral[] = [];

      this._markers.forEach((marker: IMarker) => {
        path.push(marker.position);
      })

      return [
        <Polyline
          key={ this.uuidV4() }
          options={ this.polylineOptions }
          path={ path }
        />,
      ];
    }

    return [];
  }

  private uuidV4 = (): string => {
    // tslint:disable:no-bitwise
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
    // tslint:enable:no-bitwise
  }
}
