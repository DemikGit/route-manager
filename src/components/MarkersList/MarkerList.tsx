import * as React from 'react';
import { Component } from 'react';
import { IMarker } from '../../classes/RMRoute';

import { MarkerWrapper } from '../MarkerWrapper/MarkerWrapper';
import './MarkerList.css';


export interface IMarkerListProps {
  routeMarkers: IMarker[],
}

export class MarkerList extends Component<IMarkerListProps, object> {
  public render() {
    return (
      <div className="marker-list__container">
        {
          this.props.routeMarkers.map(( marker , index ) => {
            return (
              <div
                key={ index.toString() }
                className="marker-list__marker"
              >
                <MarkerWrapper name={ marker.name } />
              </div>
            );
          })
        }
      </div>
    );
  }
}
