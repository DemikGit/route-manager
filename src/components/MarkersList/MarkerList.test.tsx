import { shallow, mount } from 'enzyme';
import 'jest-enzyme';
import * as React from 'react';
import { MarkerList } from './MarkerList';
import { IMarker } from '../../services/RouteService';

const testMarkers: IMarker[] = [
  {
    draggable: true,
    id: 'uuid',
    name: 'Test Marker',
    onDrag: (event: google.maps.MouseEvent, id: string) => {
      this.onDrag(event, id);
    },
    position: {
      lat: 0,
      lng: 0,
    },
    showInfoBox: false,
  }
];

it('renders without crashing', () => {

  const onPointDelete = jest.fn();
  const swapPoints = jest.fn();

  const wrapper = shallow(
    <MarkerList
      routeMarkers={ testMarkers }
      onPointDelete={ onPointDelete }
      swapPoints={ swapPoints }
    />
  );

  expect(wrapper).toMatchSnapshot();
});


it('change srcId if onDragStart was called', () => {
  const onPointDelete = jest.fn();
  const swapPoints = jest.fn();
  const testId = 'uuid';

  const wrapper = shallow(
    <MarkerList
      routeMarkers={ testMarkers }
      onPointDelete={ onPointDelete }
      swapPoints={ swapPoints }
    />
  );

  wrapper.instance().onDragStart(testId);

  expect(wrapper.instance().state.srcId).toEqual(testId);
  expect(wrapper).toMatchSnapshot();
});

it('call swapPoint if onDragEnter was called', () => {
  const onPointDelete = jest.fn();
  const swapPoints = jest.fn();
  const testId = 'uuid';

  const wrapper = shallow(
    <MarkerList
      routeMarkers={ testMarkers }
      onPointDelete={ onPointDelete }
      swapPoints={ swapPoints }
    />
  );

  wrapper.instance().onDragEnter(testId);

  expect(swapPoints).toBeCalled();
  expect(wrapper).toMatchSnapshot();
});
