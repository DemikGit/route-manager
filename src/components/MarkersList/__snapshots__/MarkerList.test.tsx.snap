// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call swapPoint if onDragEnter was called 1`] = `
<div
  className="marker-list__container"
>
  <div
    className="marker-list__marker"
    key="uuid"
  >
    <MarkerWrapper
      id="uuid"
      name="Test Marker"
      onDragEnter={[Function]}
      onDragStart={[Function]}
      onPointDelete={[Function]}
    />
  </div>
</div>
`;

exports[`change srcId if onDragStart was called 1`] = `
<div
  className="marker-list__container"
>
  <div
    className="marker-list__marker"
    key="uuid"
  >
    <MarkerWrapper
      id="uuid"
      name="Test Marker"
      onDragEnter={[Function]}
      onDragStart={[Function]}
      onPointDelete={[Function]}
    />
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div
  className="marker-list__container"
>
  <div
    className="marker-list__marker"
    key="uuid"
  >
    <MarkerWrapper
      id="uuid"
      name="Test Marker"
      onDragEnter={[Function]}
      onDragStart={[Function]}
      onPointDelete={[Function]}
    />
  </div>
</div>
`;
